fluent-syntax/src/serializer.rs

```diff
@@ -0,0 +1,747 @@
+    );
+    round_trip_test!(unicode_escape_sequence, "foo = { \"\\u0065\" }\n",);
+
+    // Serialize padding around comments
+
+    round_trip_test!(
+        standalone_comment_has_not_padding_when_first,
+        "# Comment A\n\nfoo = Foo\n\n# Comment B\n\nbar = Bar\n"
+    );
+    round_trip_test!(
+        group_comment_has_not_padding_when_first,
+        "## Group A\n\nfoo = Foo\n\n## Group B\n\nbar = Bar\n"
+    );
+    round_trip_test!(
+        resource_comment_has_not_padding_when_first,
+        "### Resource Comment A\n\nfoo = Foo\n\n### Resource Comment B\n\nbar = Bar\n"
+    );
+}
+
+#[cfg(test)]
+mod serialize_expression_tests {
+    use super::*;
+
+    macro_rules! expression_test {
+        ($name:ident, $input:expr) => {
+            #[test]
+            fn $name() {
+                let input_without_tabs = $input.replace("\t", "    ");
+                let src = format!("foo = {{ {} }}", input_without_tabs);
+                let resource = crate::parser::parse(src.as_str()).unwrap();
+
+                // extract the first expression from the value of the first
+                // message
+                assert_eq!(resource.body.len(), 1);
+                let first_item = &resource.body[0];
+                let message = match first_item {
+                    Entry::Message(msg) => msg,
+                    other => panic!("Expected a message but found {:#?}", other),
+                };
+                let value = message.value.as_ref().expect("The message has a value");
+                assert_eq!(value.elements.len(), 1);
+                let expr = match &value.elements[0] {
+                    PatternElement::Placeable { expression } => expression,
+                    other => panic!("Expected a single expression but found {:#?}", other),
+                };
+
+                // we've finally extracted the first expression, now we can
+                // actually serialize it and finish the test
+                let mut serializer = Serializer::new(Options::default());
+                serializer.serialize_expression(expr).unwrap();
+                let got = serializer.into_serialized_text();
+
+                assert_eq!(got, input_without_tabs);
+            }
+        };
+    }
+
+    expression_test!(string_expression, "\"str\"");
+    expression_test!(number_expression, "3");
+    expression_test!(message_reference, "msg");
+    expression_test!(external_arguemnt, "$ext");
+    expression_test!(attribute_expression, "msg.attr");
+    expression_test!(call_expression, "BUILTIN(3.14, kwarg: \"value\")");
+    expression_test!(select_expression, "$num ->\n   *[one] One\n");
+}
+
+#[cfg(test)]
+mod serialize_variant_key_tests {
+    use super::*;
+
+    macro_rules! variant_key_test {
+        ($name:ident, $input:expr => $( $keys:expr ),+ $(,)?) => {
+            #[test]
+            #[allow(unused_assignments)]
+            fn $name() {
+                let input_without_tabs = $input.replace("\t", "    ");
```

Comment:
Suggestion: I guess it's worth repeating here as a comment and elsewhere why tabs get replaced.